### PR TITLE
5823: Copying custom fields from quote to order table using sales_convert_quote in fieldset.xml

### PR DESCRIPTION
**Pull request by RomaKis**
Copy custom fields from quote table to order.
### Description
Observer created.

### Fixed Issues (if relevant)

1. magento/magento2#5823: Copying custom fields from quote to order table using sales_convert_quote in fieldset.xml

### Manual testing scenarios

1. Create a custom module that add a field to quote and sales_order table.
2. Update the field in quote table and then place an order.
3. Field from quote table must be copied to order table.

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/Sales/Observer/SalesEventQuoteSubmitBeforeObserver.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    {
+        /** @var \Magento\Quote\Api\Data\CartInterface $quote */
+        $quote = $observer->getEvent()->getQuote();
+
+        /** @var \Magento\Sales\Api\Data\OrderInterface $order */
+        $order = $observer->getEvent()->getOrder();
+
+        $fields = $this->fieldsetConfig->getFieldset('sales_convert_quote', 'global');
+
+        $methods = get_class_methods($this->order);
```
Comment (orlangur) [inline app/code/Magento/Sales/Observer/SalesEventQuoteSubmitBeforeObserver.php]: Why do we need this if `$order` is `\Magento\Sales\Api\Data\OrderInterface`?

**Comment by orlangur** [comment]
Checking on this issue and implementation approach in Slack.

File: app/code/Magento/Sales/Observer/SalesEventQuoteSubmitBeforeObserver.php (hunk 1 above)
Comment (RomaKis) [reply app/code/Magento/Sales/Observer/SalesEventQuoteSubmitBeforeObserver.php]: Yes, this is my fault. Fixed.

File: app/code/Magento/Sales/Observer/SalesEventQuoteSubmitBeforeObserver.php [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+        $fields = $this->fieldsetConfig->getFieldset('sales_convert_quote', 'global');
+
+        $methods = get_class_methods($order);
+
+        foreach ($fields as $code => $node) {
+            $targetCode = (string)$node['to_order'];
+            $targetCode = $targetCode == '*' ? $code : $targetCode;
+
+            if (!in_array($this->getMethodName($targetCode), $methods)) {
```
Comment (7ochem) [inline app/code/Magento/Sales/Observer/SalesEventQuoteSubmitBeforeObserver.php:53]: Why not use `"get" . \Magento\Framework\Api\SimpleDataObjectConverter::snakeCaseToUpperCamelCase($targetCode)`

File: app/code/Magento/Sales/Observer/SalesEventQuoteSubmitBeforeObserver.php [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    {
+        /** @var \Magento\Quote\Api\Data\CartInterface $quote */
+        $quote = $observer->getEvent()->getQuote();
+
+        /** @var \Magento\Sales\Api\Data\OrderInterface $order */
+        $order = $observer->getEvent()->getOrder();
+
+        $fields = $this->fieldsetConfig->getFieldset('sales_convert_quote', 'global');
+
+        $methods = get_class_methods($order);
```
Comment (7ochem) [inline app/code/Magento/Sales/Observer/SalesEventQuoteSubmitBeforeObserver.php:47]: The `\Magento\Framework\Api\DataObjectHelper::mergeDataObjects()` reflects methods of `\Magento\Sales\Api\Data\OrderInterface` and here methods of the implementation of that interface are used. If you have set a different preference for the `OrderInterface` and have added getter methods in that implementation, these will not be included int both copy actions: not in the one done by the DataObjectHelper and not in the one here.

You could use `\Magento\Framework\Reflection\MethodsMap::getMethodsMap(OrderInterface::class)` like is being used by the DataObjectHelper.

**Review by 7ochem — CHANGES_REQUESTED** [review]

**Comment by orlangur** [comment]
Implementation of suggested approach looks good, however current system behavior is something that does not need a fix actually. See https://github.com/magento/magento2/issues/5823#issuecomment-360987206 for details.

**Pull request closed without merging**